src/platform_impl/linux/wayland/window/shim.rs

```diff
@@ -188,6 +211,48 @@ impl WindowHandle {
         }
     }
 
+    pub fn set_user_attention(&self, request_type: Option<UserAttentionType>) {
+        let xdg_activation = match self.xdg_activation.as_ref() {
+            None => return,
+            Some(xdg_activation) => xdg_activation,
+        };
+
+        match request_type {
+            None => {
+                if let Some(token) = self.xdg_activation_token.take() {
+                    token.destroy();
+                }
+            }
+            Some(_) => {
+                // The token is already present, nothing to set.
+                if self.xdg_activation_token.borrow().is_some() {
+                    return;
+                }
+
+                let xdg_activation_token = xdg_activation.get_activation_token();
+                let surface = self.window.surface();
+                let window_id = wayland::make_wid(surface);
+                let xdg_activation = xdg_activation.clone();
+                xdg_activation_token.quick_assign(move |xdg_token, event, mut dispatch_data| {
+                    let winit_state = dispatch_data.get::<WinitState>().unwrap();
+                    if let Some(window_handle) = winit_state.window_map.get_mut(&window_id) {
+                        let surface = window_handle.window.surface();
+                        if let xdg_activation_token_v1::Event::Done { token } = event {
+                            xdg_activation.activate(token, surface);
+                            *window_handle.xdg_activation_token.borrow_mut() = None;
```

Comment:
I don't quite understand what's going on here. The `activate` call is used to raise urgency according to the protocol. Since the `xdg_activation_token` is used to "un-raise" urgency I'd thus expect this to be set to something. But here you clear it by setting it to `None`. Wouldn't that prevent people from removing urgency or does this do something different from what I'm thinking?

Reply:
The thing is that token can be used only once, thus it should be dropped right after the use, otherwise it'll result in a protocol error. Unsetting is done by compositor afaics.